src/api-extension/models/cart-action-factory.js

```diff
@@ -257,17 +257,15 @@ class CartActionFactory {
       return actions;
     }
 
-    let pos = 0;
-
-    this._cart.lineItems.forEach((item) => {
-      if (
-        cartItem.sku === item.productId &&
+    const lineItems = this._cart.lineItems.filter((item) => {
+      return (
         item.custom?.fields?.[TalonOneLineItemMetadata.effectFieldName] !== EffectType.addFreeItem
-      ) {
-        if (pos === cartItem.position) {
-          lineItem = item;
-        }
-        ++pos;
+      );
+    });
+
+    lineItems.forEach((item, i) => {
+      if (cartItem.sku === item.productId && i === cartItem.position) {
+        lineItem = item;
```

Comment:
So this is the actual "fix", right? @sebastiancarlsson

Reply:
Yes, there was a problem with this loop, for some reason that I actually didn't bother to figure out, only the first item was affected by the `setDiscountPerItem` effect, and the existing tests only had one item in the cart.. I did some more manual testing with added free items and it seems to work. Hard to test all possibilities (free items is the reason why this loop was a bit complex in the first place).